include/grpc++/completion_queue.h

```diff
@@ -75,10 +76,18 @@ class CompletionQueue {
   explicit CompletionQueue(grpc_completion_queue *take);
   ~CompletionQueue();
 
-  // Blocking read from queue.
-  // Returns true if an event was received, false if the queue is ready
-  // for destruction.
-  bool Next(void **tag, bool *ok);
+  // Tri-state return for AsyncNext: SHUTDOWN, GOT_EVENT, TIMEOUT
+  enum NextStatus {SHUTDOWN, GOT_EVENT, TIMEOUT};
+
+  // Nonblocking (until deadline) read from queue.
+  // Cannot rely on result of tag or ok if return is TIMEOUT
+  NextStatus AsyncNext(void **tag, bool *ok, gpr_timespec deadline);
```

Comment:
We've already got code using std::chrono elsewhere - which is why I made
the suggestion. I realize it's out for 2010, but the solution we choose
there will be pretty much a copy/paste when we choose it. My main point was
that it should _not_ be gpr_timespec - we've got good hygiene presently at
not exposing grpc/ headers in grpc++/, and I think that's worthwhile
sticking to.

Suggest using std::chrono now, and when we do 2010, defining some minimal
compatible C++ type for use there.

On Sun, Mar 15, 2015 at 4:45 PM Vijay Pai <notifications@github.com> wrote:

> In include/grpc++/completion_queue.h
> <https://github.com/grpc/grpc/pull/1023#discussion_r26454117>:
>
> > @@ -75,10 +76,18 @@ class CompletionQueue {
> >    explicit CompletionQueue(grpc_completion_queue *take);
> >    ~CompletionQueue();
> >
> > -  // Blocking read from queue.
> > -  // Returns true if an event was received, false if the queue is ready
> > -  // for destruction.
> > -  bool Next(void **tag, bool *ok);
> > +  // Tri-state return for AsyncNext: SHUTDOWN, GOT_EVENT, TIMEOUT
> > +  enum NextStatus {SHUTDOWN, GOT_EVENT, TIMEOUT};
> > +
> > +  // Nonblocking (until deadline) read from queue.
> > +  // Cannot rely on result of tag or ok if return is TIMEOUT
> > +  NextStatus AsyncNext(void **tag, bool *ok, gpr_timespec deadline);
>
> https://msdn.microsoft.com/en-us/library/hh874757.aspx confirms that this
> is a VS12+ thing. If our C++11 standard is based on VS10, then it's a
> no-go. I realize that that decision would affect some tests, as some tests
> are already using std::chrono.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc/pull/1023/files#r26454117>.
>